### PR TITLE
Tooltip clearing on layout changes

### PR DESCRIPTION
**Pull request by davep**
Fix for #3045. While the initial issue was about how the tooltip remains if the widget that it pertains to has been removed, or scrolled away, really the issue comes down to if *any* layout change has happened such that the widget isn't the one "under the mouse" any more (so including things like `visible` or `display` changing, or other widgets changing in size such that the target widget gets moved, etc).

The idea here is that if a `Screen._refresh_layout` takes place, a check is made to ensure that the widget under the cursor hasn't changed from the one relating to the tooltip. If it has, the tooltip is removed. To implement this, and to allow a "hook" that other people can hook into if they need to do similar things, we've created a signal that is published when a screen's layout is refreshed; the check is then done by subscribing to that signal.

This PR also ensures that a tooltip gets removed when a screen is pushed over it.

Tested lots with [with this body of code](https://github.com/davep/textual-sandbox/blob/main/tooltip_hell.py), which covers the main cases (`remove`, scroll, `visible` toggle and `display` toggle).

---
Also adds pilot testing for tooltips.

---
Also includes a small fix to `Pilot`'s mouse event faking code to ensure that `App.mouse_position` is properly updated.

### CODE REVIEW

File: src/textual/screen.py [hunk 1]
```diff
@@ -783,7 +783,9 @@ def _refresh_layout(
         if self.is_current:
             self._compositor_refresh()
 
-        if not self.app._dom_ready:
+        if self.app._dom_ready:
```
Comment (darrenburns) [inline src/textual/screen.py:786]: Does this branch run when you scroll?

Comment (darrenburns) [inline src/textual/screen.py:786]: Does this method run when you scroll?

File: src/textual/screen.py [hunk 2]
```diff
@@ -851,6 +853,26 @@ def _update_tooltip(self, widget: Widget) -> None:
             if tooltip.display and self._tooltip_widget is widget:
                 self._handle_tooltip_timer(widget)
 
+    def _maybe_clear_tooltip(self) -> None:
+        """Check if the widget under the tooltip still pertains to the tooltip.
+
+        If they differ, the tooltip will be removed.
```
Comment (darrenburns) [inline src/textual/screen.py:859]:
```suggestion
        """Check if the widget under the mouse cursor still pertains to the tooltip.

        If they differ, the tooltip will be removed.
```

Comment (darrenburns) [reply src/textual/screen.py:859]: (Not 100% sure if this suggestion is correct)

Comment (davep) [reply src/textual/screen.py:859]: Kinda both really; but yes technically it's the mouse cursor (although the intent is that it's "under" the tooltip).

File: src/textual/screen.py (hunk 1 above)
Comment (davep) [reply src/textual/screen.py:786]: In testing it very much appears to (see the test code linked to above, one of the test areas is a scrolling region).

**Comment by davep** [comment]
While this approach is working out well, and we're agreed that it's the way go to, we're [going to expand on this a little](https://github.com/Textualize/textual/issues/3045#issuecomment-1954457816).

**Review by darrenburns — CHANGES_REQUESTED** [review]
Could you add a pilot test which checks that the tooltips are now cleared from the DOM when we expect them to be?

**Comment by davep** [comment]
Curiously it looks like there are zero pilot tests for tooltips to begin with; I wonder if it's even possible (or if they simply never got added)? I'll give it a crack.

**Comment by davep** [comment]
Tests added.

**Review by darrenburns — APPROVED** [review]

**Pull request merged**